local.d/ispell.d/ispell.1
.\" -*- Mode:Text -*-
.TH ISPELL local MIT
.SH NAME
ispell \- Correct spelling for a file
.br
munchlist \- Combine suffixes in a spelling list
.br
isexpand \- Expand suffixes in a spelling list
.SH SYNOPSIS
.B ispell
[
.B \-t
|
.B \-x
|
.B \-S
|
.B \-d
file |
.B \-p
file |
.B \-w
chars ] file .....
.br
.B ispell
[
.B \-t
|
.B \-d
file |
.B \-p
file |
.B \-w
chars ]
.B \-l
.br
.B ispell
[
.B \-t
|
.B \-d
file |
.B \-p
file
] {
.B \-a
|
.B \-A
}
.br
.B ispell
[
.B \-w
chars ]
.B \-c
.br
.B ispell
.B \-v
.br
.B munchlist
[
.B \-d
file |
.B \-e
|
.B \-w
chars ]
[ files ]
.br
.B isexpand
[ files ]
.SH DESCRIPTION
.PP
.I Ispell
is fashioned after the
.I spell
program from ITS (called
.I ispell
on Twenex systems.)  The most common usage is "ispell filename".  In this
case,
.I ispell
will display each word which does not appear in the dictionary, and
allow you to change it.  If there are "near misses" in the dictionary
(words which differ by only a single letter, a missing or extra letter,
or a pair of transposed letters), then they are also displayed.  If you
think the word is correct as it stands, you can type either "Space" to
accept it this one time, or "I" to accept it and put it in your private
dictionary.  If one of the near misses is the word you want, type the
corresponding number.
(If there are more than 10 choices,
you may have to type a carriage return to complete a single-digit number).
Finally, if none of these choices is right, you
can type "R" and you will be prompted for a replacement word.
If you want to see a list of words that might be close using wildcard
characters, type "L" to lookup a word in the system dictionary.
.PP
When a misspelled word is found, it is printed at the top of the screen.
Any near misses will be printed on the following lines, and finally, two
lines containing the word are printed at the bottom of the screen.  If
your terminal can type in reverse video, the word itself is highlighted.
.PP
The
.B \-v
option causes
.I ispell
to print its current version identification on the standard output
and exit.
.PP
The
.B \-l
or "list" option to
.I ispell
is used to produce a list of misspelled words from the standard input.
.PP
The
.B \-a
option
is intended to be used from other programs through a pipe.  In this
mode,
.I ispell
expects the standard input to consist of lines containing single words.
Each word is
read, and a single line is written to the standard output.  If the word
was found in the main dictionary, or your personal dictionary, then the
line contains only a '*'.  If the word was found through suffix removal,
then the line contains a '+', a space, and the root word.  If the word
is not in the dictionary, but there are near misses, then the line
contains an '&', a space, and a list of the near misses separated by
spaces.
Also, each near miss is capitalized the same as the input
word if unless such capitalization is illegal;
in the latter case each near miss is capitalized correctly
according to the dictionary.
Finally, if the word neither appears in the dictionary, and
there are no near misses, then the line contains only a '#'.  This mode
is also suitable for interactive use when you want to figure out the
spelling of a single word.  (These characters are the same as the codes
that the real spell program uses.)
.PP
The
.B \-A
option works just like
.BR \-a ,
except that if a line begins with the string "&Include_File&", the rest
of the line is taken as the name of a file to read for further words.
Input returns to the original file when the include file is exhausted.
Inclusion may be nested up to five deep.
The key string may be changed with the environment variable
.B INCLUDE_STRING
(the ampersands, if any, must be included).
.PP
When in the
.B \-a
mode,
.I ispell
will also accept lines of single words prefixed with either a '*' or a '@'.
A line starting with '*' tells
.I ispell
to insert the word into the user's dictionary (similar to the I command).
A line starting with '@' causes
.I ispell
to accept this word in the future (similar to the A command).
.PP
The
.B \-x
option causes
.I ispell
to remove the .bak file that it normally leaves.  The .bak file contains
the pre-corrected text.  If there are file opening / writing errors,
the .bak file may be left for recovery purposes even with the -x option.
.PP
The
.B \-S
option suppresses
.IR ispell "'s"
normal behavior of sorting the list of possible replacement words.
Some people may prefer this, since it somewhat enhances the probability
that the correct word will be low-numbered.
.PP
The
.B \-t
option selects TeX/LaTeX input mode.
TeX/LaTeX mode is also automatically selected if an input file has
the extension ".tex".
In this mode, whenever a backslash ("\e") is found,
.I ispell
will skip to the next whitespace.
Thus, for example, given
.RS
\echapter {This is a Ckapter}
\ecite{SCH86}
.RE
will find "Ckapter" but will not look for SCH.
The
.B \-t
option does not recognize the TeX comment character "%".
.PP
The
.B \-d
option is used to specify an alternate hashed dictionary file,
other than the default.  If the filename does not begin with a "/",
the library directory for the default dictionary file is prefixed.
This is useful to allow dictionaries which prefer alternate British
spellings ("centre", "tyre", etc), or add lists of special-purpose
jargon and acronyms for subclasses of documents.  There are some shortcomings
in attempting to provide foreign-language dictionaries, but something
like "-dfrench" could be made to work somewhat.
The
.B \-d
option may specify
.IR /dev/null ,
in which case the dictionary is limited to the personal one.
This may be useful for certain private dictionaries.
.PP
The
.B \-p
option is used to specify an alternate personal dictionary file.
If the file name does not begin with "/", $HOME is prefixed.  Also, the
shell variable WORDLIST may be set, which renames the personal dictionary
in the same manner.  The command line overrides WORDLIST setting.  If
neither is present "ispell.words" is used.
.PP
The
.B \-w
option may be used to specify characters other than alphabetics
which may also appear in words.  For instance,
.B \-w
"&" will allow "AT&T"
to be picked up.  Underscores are useful in many technical documents.
There is an admittedly crude provision in this option for 8-bit international
characters.
Non-printing characters may be specified in the usual way by inserting a
backslash followed by the octal character code;
e.g., "\e014" for a form feed.
Alternatively, if "n" appears in the character string, the (up to)
three characters
following are a DECIMAL code 0 - 255, for the character.
For example, to include bells and form feeds in your words (an admittedly
silly thing to do, but aren't most pedagogical examples):
.PP
n007n012
.PP
Numeric digits other than the three following "n" are simply numeric
characters.  Use of "n" does not conflict with anything because actual
alphabetics have no meaning - alphabetics are already accepted.
.I Ispell
will typically be used with input from a file, meaning that preserving
parity for possible 8 bit characters from the input text is OK.  If you
specify the -l option, and actually type text from the terminal, this may
create problems if your stty settings preserve parity.
.PP
The
.B \-c
option is primarily intended for use by the
.I munchlist
shell script.
In this mode, a list of words is read from the standard input.
For each word, a list of possible root words and suffixes will be
written to the standard output.
Some of the root words will be illegal and must be filtered from the
output by other means;
the
.I munchlist
script does this.
As an example, the command "echo BOTHER | ispell -c" produces:
.PP
.RS
.nf
BOTH
BOTHE/R
BOTH/R
.fi
.RE
.PP
Unless it has been installed without the feature by your system administrator,
.I ispell
is aware of the correct capitalizations of words in the dictionary and
in your personal dictionary.
As well as recognizing words that must be capitalized (e.g., George) and
words that must be all-capitals (e.g., NASA), it can also handle words
with "unusual" capitalization (e.g., "ITCorp" or "TeX").
If a word is capitalized incorrectly, the list of possibilities will
include all acceptable capitalizations.
(More than one capitalization may be acceptable;
for example, my dictionary lists both "ITCorp" and "ITcorp".)
Normally, this feature will not cause you surprises, but there is one
circumstance you need to be aware of.
If you add a word to your dictionary that is at the beginning of a sentence
(e.g., the first word of this paragraph if "unless" were not in the
dictionary), it will be marked as "capitalization required".
A subsequent usage of this word without capitalization (e.g., the quoted word
in the previous sentence),
.I ispell
will object and suggest the capitalized version.
You must then compare the actual spellings by eye, and then type "I"
to add the un-capitalized variant to your personal dictionary.
.PP
The rules for capitalization are as follows:
.IP (1)
Any word may appear in all capitals, as in headings.
.IP (2)
Any word that is in the dictionary in all-lowercase form may appear
either in lowercase or capitalized (as at the beginning of a sentence).
.IP (3)
Any word that has "funny" capitalization (i.e., it contains both cases
and there is an uppercase character besides the first) must appear
exactly as in the dictionary, except as permitted by rule (1).
If the word is acceptable in all-lowercase, it must appear thus in a
dictionary entry.
.PP
The
.I munchlist
shell script is used to reduce the size of dictionary files,
primarily personal dictionary files.
It is also capable of combining dictionaries from various sources.
The given
.I files
are read (standard input if no arguments are given),
reduced to a minimal set of roots and suffixes that will match the
same list of words, and written to standard output.
.PP
Normally, words that are in the default dictionary are removed by
.I munchlist
during processing.
If the list is to be used with a different dictionary, the
.B \-d
option can be used to specify an alternate (hashed) dictionary file
containing words to be removed from the output list.
If a dictionary file of
.I /dev/null
is specified, no words will be removed from the output;
this is useful when munching the primary dictionary file.
.PP
The
.B \-w
option is passed on to
.IR ispell .
The
.B \-e
("efficient") option causes the script to use a slower algorithm that uses
somewhat less space in TMPDIR (normally
.IR /usr/tmp ")."
.PP
The
.I isexpand
shell script is used to expand the various suffix flags in an
.I ispell
word list.
This script can be used when looking words up in the dictionary, or
to verify that a particular suffix flag actually produces the expected
result.
.PP
It is possible to install
.I ispell
in such a way as to only support ASCII range text if desired.
.SH ENVIRONMENT
WORDLIST	Personal dictionary file name
.br
INCLUDE_STRING	Code for file inclusion under the -A option
.br
TMPDIR	Directory used for some of munchlist's temporary files
.SH FILES
/usr/public/lib/ispell.hash
.br
/usr/dict/web2		for the Lookup function
.br
$HOME/.ispell_words	user's private dictionary
.br
/usr/public/lib/isexp[1-4].sed		sed scripts for expanding suffixes
.br
/usr/public/lib/icombine	program for combining suffix flags
.SH SEE ALSO
spell(1), egrep(1), look(1), ispell(4)
.SH BUGS
It takes about five seconds for
.I ispell
to read in the hash table.
.sp
Perhaps more than ten choices should be allowed for near misses.
.sp
The hash table is stored as a quarter-megabyte array, so a PDP-11
version does not seem likely.
.sp
.I Ispell
should understand more
.I troff
syntax, and deal more intelligently with contractions.
.sp
While alternate dictionaries for foreign languages could be defined, and
the international characters included in words, rules concerning
word endings / pluralization accommodate English only.
.sp
When the
.B \-x
flag is specified,
.I ispell
will unlink any existing .bak file.
.sp
.I Munchlist
requires tremendous amounts of temporary file space for
large dictionaries.
It does respect the TMPDIR environment variable, so this space can be
redirected.
However, a lot of the temporary space it needs is for sorting, so TMPDIR
is only a partial help on systems with an uncooperative
.IR sort (1).
As a benchmark, the 15000-word
.I dict.191
takes about 1200 blocks in TMPDIR, and 2000 in
.IR sort "'s"
temporary directories.
Munching
.I dict.191
with
.I /usr/dict/words
(28000 words output)
took another 1500 blocks or so, and ran for the better part of an hour.
.SH AUTHOR
Pace Willisson (pace@mit-vax)
.br
Collected, revised, and enhanced for the Usenet by Walt Buehring.
.br
Further enhanced and debugged by
Isaac Balbin,
Stewart Clamen,
Mark Davies,
Steve Dum,
Don Kark,
Steve Kelem,
Jim Knutson,
Geoff Kuenning,
Evan Marcus,
Dave Mason,
Rob McMahon,
Bob McQueer,
David Neves,
Joe Orost,
Israel Pinkas,
Gary Puckering,
Bill Randle,
Marc Ries,
Rich Salz,
Greg Schaffer,
George Sipe,
Perry Smith,
Stefan Taxhet,
Andrew Vignaux,
James Woods,
and Ken Yap.


local.d/ispell.d/ispell.el
;;; Spelling correction interface for GNU EMACS using "ispell"

;;; Walt Buehring
;;; Texas Instruments - Computer Science Center
;;; ARPA:  Buehring%TI-CSL@CSNet-Relay
;;; UUCP:  {smu, texsun, im4u, rice} ! ti-csl ! buehring

;;; ispell-region and associated routines added by
;;; Perry Smith
;;; pedz@bobkat
;;; Tue Jan 13 20:18:02 CST 1987

;;; extensively modified by Mark Davies and Andrew Vignaux
;;; {mark,andrew}@vuwcomp
;;; Sun May 10 11:45:04 NZST 1987

;;; Depends on the ispell program snarfed from MIT-PREP in early 
;;; 1986.  The only interactive command is "ispell-word" which should be
;;; bound to M-$.  If someone writes an "ispell-region" command, 
;;; I would appreciate a copy. -- ispell-region, ispell-buffer now added.

;;; To fully install this, add this file to your GNU lisp directory and 
;;; compile it with M-X byte-compile-file.  Then add the following to the
;;; appropriate init file:

;;;  (autoload 'ispell-word "ispell"
;;;    "Check the spelling of word in buffer." t)
;;;  (global-set-key "\e$" 'ispell-word)
;;;  (autoload 'ispell-region "ispell"
;;;    "Check the spelling of region." t)
;;;  (autoload 'ispell-buffer "ispell"
;;;    "Check the spelling of buffer." t)

;;; If run on a heavily loaded system, the initial sleep time in
;;; ispell-init-process may need to be increased.

;;; No warranty expressed or implied.  All sales final.  Void where prohibited.
;;; If you don't like it, change it.

(defconst ispell-out-name " *ispell*"
  "Name of the buffer that is associated with the 'ispell' process")

(defconst ispell-temp-name " *ispell-temp*"
  "Name of the temporary buffer that 'ispell-region' uses to hold the
filtered region")

(defvar ispell-program-name "ispell"
  "Program invoked by ispell-word and ispell-region commands.")

(defvar ispell-syntax-table nil)

(if (null ispell-syntax-table)
    ;; The following assumes that the standard-syntax-table
    ;; is static.  If you add words with funky characters
    ;; to your dictionary, the following may have to change.
    (progn
      (setq ispell-syntax-table (make-syntax-table))
      ;; Make certain characters word constituents
      ;; (modify-syntax-entry ?' "w   " ispell-syntax-table)
      ;; (modify-syntax-entry ?- "w   " ispell-syntax-table)
      ;; Get rid on existing word syntax on certain characters 
      (modify-syntax-entry ?0 ".   " ispell-syntax-table)
      (modify-syntax-entry ?1 ".   " ispell-syntax-table)
      (modify-syntax-entry ?2 ".   " ispell-syntax-table)
      (modify-syntax-entry ?3 ".   " ispell-syntax-table)
      (modify-syntax-entry ?4 ".   " ispell-syntax-table)
      (modify-syntax-entry ?5 ".   " ispell-syntax-table)
      (modify-syntax-entry ?6 ".   " ispell-syntax-table)
      (modify-syntax-entry ?7 ".   " ispell-syntax-table)
      (modify-syntax-entry ?8 ".   " ispell-syntax-table)
      (modify-syntax-entry ?9 ".   " ispell-syntax-table)
      (modify-syntax-entry ?$ ".   " ispell-syntax-table)
      (modify-syntax-entry ?% ".   " ispell-syntax-table)))


(defun ispell-word (&optional quietly)
  "Check spelling of word at or before dot.
If word not found in dictionary, display possible corrections in a window 
and let user select."
  (interactive)
  (let* ((current-syntax (syntax-table))
	 start end word poss replace)
    (unwind-protect
	(save-excursion
	  ;; Ensure syntax table is reasonable 
	  (set-syntax-table ispell-syntax-table)
	  ;; Move backward for word if not already on one.
	  (if (not (looking-at "\\w"))
	      (re-search-backward "\\w" (point-min) 'stay))
	  ;; Move to start of word
	  (re-search-backward "\\W" (point-min) 'stay)
	  ;; Find start and end of word
	  (or (re-search-forward "\\w+" nil t)
	      (error "No word to check."))
	  (setq start (match-beginning 0)
		end (match-end 0)
		word (buffer-substring start end)))
      (set-syntax-table current-syntax))
    (ispell-init-process)		; erases ispell output buffer
    (or quietly (message "Checking spelling of %s..." (upcase word)))
    (save-excursion
      (set-buffer ispell-out-name)
      (send-string ispell-process (concat word "\n"))
      ;; wait until we have a complete line
      (while (progn
	       (goto-char (point-max))
	       (/= (preceding-char) ?\n))
	(accept-process-output ispell-process))
      (goto-char (point-min))
      (setq poss (ispell-parse-output
		  (buffer-substring (point) 
				    (progn (end-of-line) (point))))))
    (cond ((eq poss t)
	   (or quietly (message "Found %s" (upcase word))))
	  ((stringp poss)
	   (or quietly (message "Found it because of %s" (upcase poss))))
	  ((null poss)
	   (or quietly (message "Could Not Find %s" (upcase word))))
	  (t (setq replace (ispell-choose poss word))
	     (if replace
		 (progn
		    (goto-char end)
		    (delete-region start end)
		    (insert-string replace)))))
    poss))


(defun ispell-choose (choices word)
  "Display possible corrections from list CHOICES.  Return chosen word
if one is chosen; Return nil to keep word"
  (unwind-protect 
      (save-window-excursion
	(let ((count 0)
	      (line 2)
	      (words choices)
	      (window-min-height 2)
	      char num result)
	  (save-excursion
	    (set-buffer (get-buffer-create "*Choices*")) (erase-buffer)
	    (setq mode-line-format "--  %b  --")
	    (while words
	      (if (<= (+ 7 (current-column) (length (car words)))
		      (window-width))
		  nil
		(insert "\n")
		(setq line (1+ line)))
	      (insert "(" (+ count ?0) ") " (car words) "  ")
	      (setq words (cdr words)
		    count (1+ count))))
	  (overlay-window line)
	  (switch-to-buffer "*Choices*")
	  (select-window (next-window))
	  (while (eq t
		     (setq result
			   (progn
			     (message "Enter letter to replace word;  Space to leave unchanged")
			     (setq char (read-char))
			     (setq num (- char ?0))
			     (cond ((= char ? ) nil)
				   ((= char ?i)
				    (send-string ispell-process
						 (concat "*" word "\n"))
				    (if (get-buffer ispell-temp-name)
					(save-excursion
					  (set-buffer ispell-temp-name)
					  (save-excursion
					    (replace-regexp (concat "^" word "$")
							    (concat "+" word)))))
				    nil)
				   ((= char ?a)
				    (send-string ispell-process
						(concat "@" word "\n"))
				    (if (get-buffer ispell-temp-name)
					(save-excursion
					  (set-buffer ispell-temp-name)
					  (save-excursion
					    (replace-regexp (concat "^" word "$")
							    (concat "+" word)))))
				    nil)
				   ((= char ?r) (read-string "Replacement: " nil))
				   ((and (>= num 0) (< num count)) (nth num choices))
				   ((= char ?\C-r)
					; Note: does not reset syntax table
				    (save-excursion (recursive-edit)) t) ; dangerous
				   ((= char ?\C-z)
				    (suspend-emacs) t)
				   ((= char help-char)
				    (message "[r]eplace; [a]ccept for session; [i] accept and add to private dictionary")
				    (sit-for 3) t)
				   (t (ding) t))))))
	  result))
    ;; Protected forms...
    (bury-buffer "*Choices*")))


(defun overlay-window (height)
  "Create a (usually small) window with HEIGHT lines and avoid
recentering."
  (save-excursion
    (let ((oldot (save-excursion (beginning-of-line) (dot)))
	  (top (save-excursion (move-to-window-line height) (dot)))
	  newin)
      (if (< oldot top) (setq top oldot))
      (setq newin (split-window-vertically height))
      (set-window-start newin top))))


(defvar ispell-process nil
  "Holds the process object for 'ispell'")

(defun ispell-parse-output (output)
"Parse the OUTPUT string of 'ispell' and return either t for an exact
match, a string containing the root word for a match via suffix
removal, a list of possible correct spellings, or nil for a complete
miss."
  (cond
   ((string= output "*") t)
   ((string= output "#") nil)
   ((string= (substring output 0 1) "+")
    (substring output 2))
   (t
    (let ((choice-list '()))
      (while (not (string= output ""))
	(let* ((start (string-match "[A-z]" output))
	       (end (string-match " \\|$" output start)))
	  (if start
	      (setq choice-list (cons (substring output start end)
				      choice-list)))
	  (setq output (substring output (1+ end)))))
      choice-list))))


(defun ispell-init-process ()
  "Check status of 'ispell' process and start if necessary."
  (if (and ispell-process
	   (eq (process-status ispell-process) 'run))
      (save-excursion
	(set-buffer ispell-out-name)
	(erase-buffer))
    (message "Starting new ispell process...")
    (and (get-buffer ispell-out-name) (kill-buffer ispell-out-name))
    (setq ispell-process (start-process "ispell" ispell-out-name
					ispell-program-name "-A"))
    (process-kill-without-query ispell-process)
    (sit-for 3)))

(defvar ispell-filter-hook "/usr/bin/tr"
  "Filter to pass a region through before sending it to ispell.
Must produce output one word per line.  Typically this is set to tr,
deroff, detex, etc.")
(make-variable-buffer-local 'ispell-filter-hook)

(defvar ispell-filter-hook-args '("-cs" "A-Za-z" "\n")
  "Argument LIST to pass to ispell-filter-hook")
(make-variable-buffer-local 'ispell-filter-hook-args)

; This routine has certain limitations brought about by the filter
; hook.  For example, deroff will take ``\fBcat\fR'' and spit out
; ``cat''.  This is hard to search for since word-search-forward will
; not match at all and search-forward for ``cat'' will match
; ``concatenate'' if it happens to occur before.
; `ispell-region' filters the region into `*ispell-temp*', writes the
; buffer to a temporary file, and sends a ``&Include_File&foobar''
; string to the ispell process which is writing into `*ispell*'.
; `ispell-region' then searches `*ispell*' for a spelling error (`#' or
; `&'), checks the `*ispell-temp*' buffer for the misspelled word and
; then skips forward `count' words (the number of correct lines in
; `*ispell*') in the region.  It then searches for the misspelled
; word.  This is not a foolproof heuristic but it is fast and works
; most of the time.

(defun ispell-region (start end)
  "Check a region for spelling errors interactively.  The variable
which should be buffer or mode specific ispell-filter-hook is called
to filter out text processing commands."
  (interactive "r")
  (let ((this-buf (current-buffer))
	(spell-file (make-temp-name "/usr/tmp/is"))
	(spell-buf (get-buffer-create ispell-temp-name))
	(current-syntax (syntax-table))
	(tracker 1)
	word poss replace endbound ispell-out)
    (ispell-init-process)
    (setq ispell-out (get-buffer ispell-out-name))
    (unwind-protect
	(save-excursion
	  (message "Prefrobnicating...")
	  (sit-for 0)
	  (set-syntax-table ispell-syntax-table)
	  (set-buffer spell-buf)
	  (erase-buffer)
	  (set-buffer this-buf)
	  (apply 'call-process-region 
		 (append (list start end ispell-filter-hook nil spell-buf nil)
			 ispell-filter-hook-args))
	  (goto-char start)
	  (set-buffer spell-buf)
	  (and (/= (preceding-char) ?\n) ; couple of hacks for tr
	       (insert "\n"))
	  (goto-char (point-min))
	  (while (= (following-char) ?\n)
	    (delete-char 1))
	  (write-region (point-min) (point-max) spell-file nil 1)
	  (send-string ispell-process 
		       (concat "&Include_File&" spell-file "\n"))
	  (message "Looking for a misspelled word") (sit-for 0)
	  (while (not (eobp))
	    (set-buffer ispell-out)
	    (goto-char (point-max))
	    (beginning-of-line)
	    (setq endbound (point))
	    (goto-char tracker)
	    (if (prog1
		    (not (re-search-forward "^[#&]" endbound 1))
		  (beginning-of-line)
		  (setq count (count-lines tracker (point))
			tracker (point))
		  (set-buffer spell-buf)
		  (forward-line count))
		(accept-process-output)	; give it some time to get something
	      (setq word (buffer-substring (point)
					   (progn (end-of-line) (point))))
	      (forward-char 1)
	      (set-buffer ispell-out) ; (goto-char tracker)
	      (setq poss (ispell-parse-output
			  (buffer-substring (point) 
					    (progn (end-of-line) (point)))))
	      (forward-char 1)
	      (setq tracker (point))
	      (set-buffer this-buf)
	      (re-search-forward "\\W*\\(\\w+\\)" nil t (1- count)) ; get close
	      (if (string= "+" (substring word 0 1))
		  (search-forward (substring word 1) nil t)
		(if (search-forward word nil t)
		    (progn
		      (sit-for 0)
		      (setq replace (ispell-choose poss word))
		      (if replace
			  (replace-match replace)))
		  (message "Can not find %s in original text -- Any key to continue" word)
		  (and (= ?\C-z (read-char)) (suspend-emacs)))
		(message "Looking for a misspelled word") (sit-for 0))
	      (set-buffer spell-buf))))
      (message "")
      (set-syntax-table current-syntax)
      (and (file-exists-p spell-file)
	   (delete-file spell-file)))))

(defun ispell-buffer () 
  "Check the current buffer for spelling errors interactively.  The variable
which should be buffer or mode specific ispell-filter-hook is called to
filter out text processing commands."
  (interactive)
  (ispell-region (point-min) (point-max)))
